contracts/fungible-token-full.aes

```diff
@@ -72,14 +71,20 @@ contract FungibleTokenFull =
     require(String.length(symbol) >= 1, "STRING_TOO_SHORT_SYMBOL")
     // If the provided value for `decimals` is negative abort the execution
     require_non_negative_value(decimals)
-    // If negative initial owner balance is passed, abort the execution
-    let initial_supply = Option.default(0, initial_owner_balance)
-    require_non_negative_value(initial_supply)
 
     let owner = Call.caller
+    let (initial_supply, initial_balances) = switch(initial_owner_balance)
+      Some(initial_balance) =>
+        // If negative initial owner balance is passed, abort the execution
+        require_non_negative_value(initial_balance)
+        // Emit Mint event to signal token supply has been created
+        Chain.event(Mint(owner, initial_balance))
+        (initial_balance, { [owner] = initial_balance })
```

Comment:
Can we replace this by a call to a private function `internal_mint` proposed implementation:

```sophia
  stateful function internal_mint(account: address, value: int) =
    require_non_negative_value(value)
    put(state{ total_supply = state.total_supply + value, balances[account = 0] @ b = b + value })
    Chain.event(Mint(account, value))
```

So that it can be used both in the public [mint entrypoint ](which is to also be used by the mint entrypoint here:
https://github.com/aeternity/aeternity-fungible-token/blob/33b4b62980122e1fcef7f4f681b7f2517e709602/contracts/fungible-token-full.aes#L193-L197) like this instead:

```sophia
  stateful entrypoint mint(account: address, value: int) =
    internal_mint(account, value)
```

and also in the `init` function in the sam

Reply:
I agree, for the full example this works, for the simple token it has to be its own logic

Reply:
Actually maybe this won't be so good.

1. having the `require_owner` excluded from the internal function would maybe encourage incorrect behaviour in using it
2. state is not defined yet in init, thus the same code wouldn't work to be called from init, referencing state

Reply:
let me know once you found consensus around that. if it's not possible like @mradkov proposed we need to stick to this route, right? I'd approve then :)

Reply:
> Actually maybe this won't be so good.
> 
> 1. having the `require_owner` excluded from the internal function would maybe encourage incorrect behaviour in using it
> 2. state is not defined yet in init, thus the same code wouldn't work to be called from init, referencing state

In NFT drops usually the users are the ones to call the `mint` function, thus I suggested the removal of the default `require_owner` there.

Reply:
@mradkov this is about the fungible tokens 🤔 I think anyway if some special logic is required it needs to be adapted by the developers - even for NFTs. do you agree?

Reply:
yeah, for the reference implementation this is more safe though

Reply:
I would argue to keep it as it is